File: js/src/graphemes.js
// Grapheme clusters — user-perceived characters that span several code points.
//
// Failure mode #10 (grapheme-cluster-split): a single thing a reader sees as
// "one character" is often several code points glued together. A ZWJ emoji
// family (👨‍👩‍👧 is man + ZWJ + woman + ZWJ + girl), a skin-tone emoji (👍🏽 is
// thumb + a U+1F3FB..U+1F3FF modifier), a flag (🇯🇵 is two regional indicators),
// or a Japanese ideographic variation sequence (葛 + U+E0100 selects a specific
// official-register glyph for a name kanji) are each one grapheme.
//
// Truncating, reversing, or measuring such text by code point — even with the
// correct code-point slice that fixes the byte (#2) and surrogate (#8) bugs —
// still cuts these apart: a family loses members, a skin tone is stripped (which
// reads as changing the person), a flag becomes two letters, and a name kanji
// silently changes glyph. The fix is to segment by grapheme cluster. In a
// browser or Node, Intl.Segmenter({ granularity: "grapheme" }) does this; the
// minimal segmenter below covers the joiners these fixtures exercise without a
// dependency.

const ZWJ = 0x200d;

// Code points that attach to the previous one (Grapheme_Cluster_Break = Extend
// or SpacingMark, narrowed to what these fixtures use): combining marks across
// scripts, variation selectors (incl. the supplementary IVS block), and emoji
// skin-tone modifiers. A small, dependency-free table, not the full UCD.
function isExtend(cp) {
  return (
    (cp >= 0x0300 && cp <= 0x036f) || // combining diacritical marks (Latin)
    (cp >= 0x0591 && cp <= 0x05bd) || // Hebrew points
    (cp >= 0x064b && cp <= 0x065f) || // Arabic harakat
    (cp >= 0xfe00 && cp <= 0xfe0f) || // variation selectors
    (cp >= 0x1f3fb && cp <= 0x1f3ff) || // emoji skin-tone modifiers
    (cp >= 0xe0100 && cp <= 0xe01ef) // variation selectors supplement (IVS)
  );
}

const isRegionalIndicator = (cp) => cp >= 0x1f1e6 && cp <= 0x1f1ff;

/**
 * Correct: split a string into grapheme clusters. Code points are kept together
 * when joined by ZWJ, when one is an Extend mark / variation selector / skin
 * tone, or when two regional indicators pair into a flag.
 */
export function graphemes(str) {
  const out = [];
  let cur = "";
  let prevZWJ = false;
  let ri = 0; // regional indicators in the current cluster
  for (const ch of str) {
    const cp = ch.codePointAt(0);
    if (cur === "") {
      cur = ch;
      prevZWJ = cp === ZWJ;
      ri = isRegionalIndicator(cp) ? 1 : 0;
      continue;
    }
    if (cp === ZWJ) {
      cur += ch;
      prevZWJ = true;
      continue;
    }
    if (prevZWJ) {
      cur += ch;
      prevZWJ = false;
      ri = isRegionalIndicator(cp) ? 1 : 0;
      continue;
    }
    if (isExtend(cp)) {
      cur += ch;
      continue;
    }
    if (isRegionalIndicator(cp) && ri === 1) {
      cur += ch;
      ri = 2;
      continue;
    }
    out.push(cur);
    cur = ch;
    prevZWJ = cp === ZWJ;
    ri = isRegionalIndicator(cp) ? 1 : 0;
  }
  if (cur !== "") out.push(cur);
  return out;
}

/** Count of grapheme clusters (what a reader would call "characters"). */
export function graphemeLength(str) {
  return graphemes(str).length;
}

/** Correct: take the first n grapheme clusters, never splitting one. */
export function graphemeSlice(str, n) {
  return graphemes(str).slice(0, n).join("");
}

/** Buggy reference: take the first n code points, splitting joined clusters. */
export function codePointSlice(str, n) {
  return Array.from(str).slice(0, n).join("");
}

File: go/graphemes.go
package cjkfixtures

import "strings"

// Grapheme clusters (failure mode #10, grapheme-cluster-split). One thing a
// reader sees as a single character is often several code points glued together:
// a ZWJ emoji family (👨‍👩‍👧 is man + ZWJ + woman + ZWJ + girl), a skin-tone emoji
// (👍 + a U+1F3FB..U+1F3FF modifier), a flag (🇯🇵 is two regional indicators), or a
// Japanese ideographic variation sequence (a name kanji + U+E0100 selecting an
// official-register glyph). Truncating or measuring by code point — even with
// the correct rune slice that fixes #2 and #8 — still cuts these apart. The fix
// is to segment by grapheme cluster.
//
// The Go standard library has no grapheme segmenter (that lives in
// golang.org/x/text). The minimal segmenter below covers the joiners these
// fixtures exercise (ZWJ, variation selectors, skin tones, regional-indicator
// pairs, combining marks) without a dependency; it is not the full UAX #29.

const zwj = 0x200d

// isGraphemeExtend reports whether a rune attaches to the previous one
// (Grapheme_Cluster_Break = Extend, narrowed to what these fixtures use).
func isGraphemeExtend(r rune) bool {
	switch {
	case r >= 0x0300 && r <= 0x036f: // combining diacritical marks (Latin)
		return true
	case r >= 0x0591 && r <= 0x05bd: // Hebrew points
		return true
	case r >= 0x064b && r <= 0x065f: // Arabic harakat
		return true
	case r >= 0xfe00 && r <= 0xfe0f: // variation selectors
		return true
	case r >= 0x1f3fb && r <= 0x1f3ff: // emoji skin-tone modifiers
		return true
	case r >= 0xe0100 && r <= 0xe01ef: // variation selectors supplement (IVS)
		return true
	}
	return false
}

func isRegionalIndicator(r rune) bool { return r >= 0x1f1e6 && r <= 0x1f1ff }

// Graphemes splits a string into grapheme clusters: code points are kept
// together when joined by ZWJ, when one is an Extend mark, or when two regional
// indicators pair into a flag.
func Graphemes(s string) []string {
	out := []string{}
	cur := []rune{}
	prevZWJ := false
	ri := 0 // regional indicators in the current cluster
	startCluster := func(r rune) {
		cur = []rune{r}
		prevZWJ = r == zwj
		if isRegionalIndicator(r) {
			ri = 1
		} else {
			ri = 0
		}
	}
	for _, r := range s {
		switch {
		case len(cur) == 0:
			startCluster(r)
		case r == zwj:
			cur = append(cur, r)
			prevZWJ = true
		case prevZWJ:
			cur = append(cur, r)
			prevZWJ = false
			if isRegionalIndicator(r) {
				ri = 1
			} else {
				ri = 0
			}
		case isGraphemeExtend(r):
			cur = append(cur, r)
		case isRegionalIndicator(r) && ri == 1:
			cur = append(cur, r)
			ri = 2
		default:
			out = append(out, string(cur))
			startCluster(r)
		}
	}
	if len(cur) > 0 {
		out = append(out, string(cur))
	}
	return out
}

// GraphemeLength counts grapheme clusters (what a reader calls "characters").
func GraphemeLength(s string) int {
	return len(Graphemes(s))
}

// GraphemeSlice takes the first n grapheme clusters, never splitting one.
func GraphemeSlice(s string, n int) string {
	g := Graphemes(s)
	if n > len(g) {
		n = len(g)
	}
	return strings.Join(g[:n], "")
}
